src/System.Net.Sockets/src/System/Net/Sockets/IOControlKeepAlive.Windows.cs

```diff
@@ -0,0 +1,118 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace System.Net.Sockets
+{
+    // Allow hiding keep-alive time and interval handling behind *SocketOption
+    // on Windows < 10 v1709 that only supports set via IOControl
+    internal class IOControlKeepAlive
+    {
+        private const uint WindowsDefaultTimeMs = 7200000u;
+        private const uint WindowsDefaultIntervalMs = 1000u;
+        private static readonly bool s_supportsKeepAliveViaSocketOption = SupportsKeepAliveViaSocketOption();
+        private static readonly ConditionalWeakTable<SafeSocketHandle, IOControlKeepAlive> s_socketKeepAliveTable = new ConditionalWeakTable<SafeSocketHandle, IOControlKeepAlive>();
+        [ThreadStatic]
+        private static byte[] s_keepAliveValuesBuffer;
+
+        private uint _timeMs = WindowsDefaultTimeMs;
+        private uint _intervalMs = WindowsDefaultIntervalMs;
+
+        public static bool IsNeeded
+        {
+            get
+            {
+                return !s_supportsKeepAliveViaSocketOption;
+            }
+        }
+
+        public static SocketError Get(SafeSocketHandle handle, SocketOptionName optionName, byte[] optionValueSeconds, ref int optionLength)
+        {
+            Debug.Assert(optionValueSeconds != null);
+            Debug.Assert(optionValueSeconds.Length == sizeof(int));
```

Comment:
Couldn't a null optionValueSeconds make it in here?  I don't think these asserts are valid.

Reply:
Useless to say you are right (as usual)... late night coding doesn't help 😞